documentation/specs/src/base-ledger/replay-protection.md

```diff
@@ -0,0 +1,103 @@
+
+```rust
+pub struct Tx {
+    pub code: Vec<u8>,
+    pub data: Option<Vec<u8>>,
+    pub timestamp: DateTimeUtc,
+    pub tx_counter: Option<(Address, u64)>
+}
+```
+
+The check would run in `process_proposal` and the update in `finalize_block`, just like for the wrapper transaction. This implementation, though, shows two drawbacks:
+
+- it implies the need for an hard fork in case of a modification of the replay protection mechanism
+- it's not clear who's the source of the inner transaction from the outside, as that depends on the specific code of the transaction itself. We could use specific whitelisted txs set to define when it requires a counter (would not work for future programmable transactions), but still, we have no way to define which address should be targeted for replay protection (**blocking issue**)
+
+### VPs and Txs
+
+To implement replay protection for the inner transaction we will need to update all the VPs checking the transaction's signature to include the check on the transaction counter: at the moment the `vp_user` validity predicate is the only one to update.
+
+In addition, all the transactions involving `SignedTxData` should increment the counter.
+
+## Single counter
+
+The mechanisms to prevent replay attacks for both transactions (wrapper and inner) will share the same `tx_counter` key in storage.
+
+This could be an issue when the signer of the `WrapperTx` is the same of the `InnerTx` (default behaviour of the client): in this case, if both transactions expect the same counter in storage, the wrapper transaction will pass validation but the inner one will fail. To cope with this, the client will be responsible for producing a valid `WrapperTx` in which `tx_counter` will be set to the current value of the counter in storage, call it `storage_ctr`, while the inner transaction will have `storage_ctr + 1` in its data.
+
+## Storage Interface
+
+Replay protection will require interaction with the storage from both the protocol and Wasm. To do so we can take advantage of the `StorageRead` and `StorageWrite` traits to work with a single interface.
+
+## Batching
+
+The implementation proposed in this document doesn't support batching of multiple transactions from a same address in a single block. More specifically, the transactions will all succeed only if they are executed in the intended order, but the order in which transactions will be included in the block by the proposer is not guaranteed. An out of order execution of multiple transactions would lead to the failure of some of them (in the worst case, the failure of all of them but one, in the best case, the failure of only one). This problem will be amplified by the introduction of Ferveo for DKG which will be able to reorder transactions.
```

Comment:
well, the outer tx nonce can be incremented in the submitted transactions, right? this at least gives the proposer an incentive to order them correctly since they receive more fees, right?

Reply:
Yes, the outer counter needs to be incremented every time. If the order of transactions in the mempool of the proposer is the intended one, all the transactions will be executed correctly. The issue is when the order is off: at the moment, the protocol doesn't have a way to reorder the txs based on this counter (we could implement it). This of course works assuming that we are going to charge the fee (or at least part of it) for the inclusion in the block only, regardless of the transaction being correctly executed, otherwise, the proposer would just have an incentive to add them to the block without worrying about the order.
But if we will randomize the tx order for DKG then we are stuck with this problem.  
Another scenario is if, for any reason, a transaction doesn't get to the proposer's mempool: in this case, all the transactions with a tx_counter greater than the missing one will fail, even if placed in the correct (ascending) order.

Reply:
the protocol doesn't need to reorder the transactions as long as it is incentive compatible for the proposer to order them correctly

we don't need to worry about the DKG for now (perhaps such a counter would have to be outside the encrypted inner transaction if the user wants a particular execution order to be enforced)

Reply:
for the latter case, we should probably prevent transactions with counter different by more than 1 from executing / paying any fee at all